File: src/main/java/net/Chidoziealways/everythingjapanese/block/custom/HellPortalBlock.java
package net.Chidoziealways.everythingjapanese.block.custom;

import com.mojang.serialization.MapCodec;
import net.Chidoziealways.everythingjapanese.worldgen.dimension.ModDimensions;
import net.minecraft.BlockUtil;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.resources.ResourceKey;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.EntityDimensions;
import net.minecraft.world.entity.InsideBlockEffectApplier;
import net.minecraft.world.entity.Relative;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.GameRules;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.LevelReader;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.Portal;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.level.block.state.properties.EnumProperty;
import net.minecraft.world.level.border.WorldBorder;
import net.minecraft.world.level.dimension.DimensionType;
import net.minecraft.world.level.portal.PortalShape;
import net.minecraft.world.level.portal.TeleportTransition;
import net.minecraft.world.phys.Vec3;
import net.minecraft.world.phys.shapes.Shapes;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;
import java.util.Optional;

public class HellPortalBlock extends Block implements Portal{
    private static final Logger LOGGER = LoggerFactory.getLogger(HellPortalBlock.class);
    public static final MapCodec<HellPortalBlock> CODEC = simpleCodec(HellPortalBlock::new);
    public static final EnumProperty<Direction.Axis> AXIS = BlockStateProperties.HORIZONTAL_AXIS;
    private static final Map<Direction.Axis, VoxelShape> SHAPES = Shapes.rotateHorizontalAxis(Block.column(4.0, 16.0, 0.0, 16.0));

    @Override
    protected MapCodec<? extends Block> codec() {
        return CODEC;
    }

    public HellPortalBlock(Properties properties) {
        super(properties);
        this.registerDefaultState(this.stateDefinition.any().setValue(AXIS, Direction.Axis.X));
    }

    @Override
    protected VoxelShape getEntityInsideCollisionShape(BlockState p_396779_, BlockGetter p_395715_, BlockPos p_396107_, Entity p_395206_) {
        return p_396779_.getShape(p_395715_, p_396107_);
    }

    @Override
    protected void entityInside(BlockState p_54915_, Level p_54916_, BlockPos p_54917_, Entity p_54918_, InsideBlockEffectApplier p_392916_) {
        if (p_54918_.canUsePortal(false)) {
            p_54918_.setAsInsidePortal(this, p_54917_);
        }
    }

    @Override
    public int getPortalTransitionTime(ServerLevel p_342064_, Entity p_344634_) {
        return p_344634_ instanceof Player player
                ? Math.max(0, p_342064_.getGameRules().getInt(player.getAbilities().invulnerable ? GameRules.RULE_PLAYERS_NETHER_PORTAL_CREATIVE_DELAY : GameRules.RULE_PLAYERS_NETHER_PORTAL_DEFAULT_DELAY))
                : 0;
    }

    @Override
    public @Nullable TeleportTransition getPortalDestination(ServerLevel pLevel, Entity pEntity, BlockPos pPos) {
        ResourceKey<Level> resourcekey = pLevel.dimension() == ModDimensions.HELL_LEVEL_KEY ? Level.OVERWORLD : ModDimensions.HELL_LEVEL_KEY;
        ServerLevel serverlevel = pLevel.getServer().getLevel(resourcekey);
        if (serverlevel == null) {
            return null;
        } else {
            boolean flag = serverlevel.dimension() == ModDimensions.HELL_LEVEL_KEY;
            WorldBorder worldborder = serverlevel.getWorldBorder();
            double d0 = DimensionType.getTeleportationScale(pLevel.dimensionType(), serverlevel.dimensionType());
            BlockPos blockpos = worldborder.clampToBounds(pEntity.getX() * d0, pEntity.getY(), pEntity.getZ() * d0);
            return this.getExitPortal(serverlevel, pEntity, pPos, blockpos, flag, worldborder);
        }
    }

    @javax.annotation.Nullable
    private TeleportTransition getExitPortal(
            ServerLevel pLevel, Entity pEntity, BlockPos pPos, BlockPos pExitPos, boolean pIsNether, WorldBorder pWorldBorder
    ) {
        Optional<BlockPos> optional = pLevel.getPortalForcer().findClosestPortalPosition(pExitPos, pIsNether, pWorldBorder);
        BlockUtil.FoundRectangle blockutil$foundrectangle;
        TeleportTransition.PostTeleportTransition teleporttransition$postteleporttransition;
        if (optional.isPresent()) {
            BlockPos blockpos = optional.get();
            BlockState blockstate = pLevel.getBlockState(blockpos);
            blockutil$foundrectangle = BlockUtil.getLargestRectangleAround(
                    blockpos, blockstate.getValue(BlockStateProperties.HORIZONTAL_AXIS), 21, Direction.Axis.Y, 21, p_343533_ -> pLevel.getBlockState(p_343533_) == blockstate
            );
            teleporttransition$postteleporttransition = TeleportTransition.PLAY_PORTAL_SOUND.then(p_343530_ -> p_343530_.placePortalTicket(blockpos));
        } else {
            Direction.Axis direction$axis = pEntity.level().getBlockState(pPos).getOptionalValue(AXIS).orElse(Direction.Axis.X);
            Optional<BlockUtil.FoundRectangle> optional1 = pLevel.getPortalForcer().createPortal(pExitPos, direction$axis);
            if (optional1.isEmpty()) {
                LOGGER.error("Unable to create a portal, likely target out of worldborder");
                return null;
            }

            blockutil$foundrectangle = optional1.get();
            teleporttransition$postteleporttransition = TeleportTransition.PLAY_PORTAL_SOUND.then(TeleportTransition.PLACE_PORTAL_TICKET);
        }

        return getDimensionTransitionFromExit(pEntity, pPos, blockutil$foundrectangle, pLevel, teleporttransition$postteleporttransition);
    }

    private static TeleportTransition getDimensionTransitionFromExit(
            Entity pEntity, BlockPos pPos, BlockUtil.FoundRectangle pRectangle, ServerLevel pLevel, TeleportTransition.PostTeleportTransition pPostTeleportTransition
    ) {
        BlockState blockstate = pEntity.level().getBlockState(pPos);
        Direction.Axis direction$axis;
        Vec3 vec3;
        if (blockstate.hasProperty(BlockStateProperties.HORIZONTAL_AXIS)) {
            direction$axis = blockstate.getValue(BlockStateProperties.HORIZONTAL_AXIS);
            BlockUtil.FoundRectangle blockutil$foundrectangle = BlockUtil.getLargestRectangleAround(
                    pPos, direction$axis, 21, Direction.Axis.Y, 21, p_342174_ -> pEntity.level().getBlockState(p_342174_) == blockstate
            );
            vec3 = pEntity.getRelativePortalPosition(direction$axis, blockutil$foundrectangle);
        } else {
            direction$axis = Direction.Axis.X;
            vec3 = new Vec3(0.5, 0.0, 0.0);
        }

        return createDimensionTransition(pLevel, pRectangle, direction$axis, vec3, pEntity, pPostTeleportTransition);
    }

    private static TeleportTransition createDimensionTransition(
            ServerLevel pLevel,
            BlockUtil.FoundRectangle pRectangle,
            Direction.Axis pAxis,
            Vec3 pOffset,
            Entity pEntity,
            TeleportTransition.PostTeleportTransition pPostTeleportTransition
    ) {
        BlockPos blockpos = pRectangle.minCorner;
        BlockState blockstate = pLevel.getBlockState(blockpos);
        Direction.Axis direction$axis = blockstate.getOptionalValue(BlockStateProperties.HORIZONTAL_AXIS).orElse(Direction.Axis.X);
        double d0 = pRectangle.axis1Size;
        double d1 = pRectangle.axis2Size;
        EntityDimensions entitydimensions = pEntity.getDimensions(pEntity.getPose());
        int i = pAxis == direction$axis ? 0 : 90;
        double d2 = entitydimensions.width() / 2.0 + (d0 - entitydimensions.width()) * pOffset.x();
        double d3 = (d1 - entitydimensions.height()) * pOffset.y();
        double d4 = 0.5 + pOffset.z();
        boolean flag = direction$axis == Direction.Axis.X;
        Vec3 vec3 = new Vec3(blockpos.getX() + (flag ? d2 : d4), blockpos.getY() + d3, blockpos.getZ() + (flag ? d4 : d2));
        Vec3 vec31 = PortalShape.findCollisionFreePosition(vec3, pLevel, pEntity, entitydimensions);
        return new TeleportTransition(pLevel, vec31, Vec3.ZERO, i, 0.0F, Relative.union(Relative.DELTA, Relative.ROTATION), pPostTeleportTransition);
    }

    @Override
    public Portal.Transition getLocalTransition() {
        return Portal.Transition.CONFUSION;
    }

    @Override
    protected ItemStack getCloneItemStack(LevelReader pLevel, BlockPos pPos, BlockState pState, boolean pIncludeData) {
        return ItemStack.EMPTY;
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> pBuilder) {
        pBuilder.add(AXIS);
    }
}

File: src/main/kotlin/net/Chidoziealways/everythingjapanese/block/custom/HellPortalBlock.kt
package net.Chidoziealways.everythingjapanese.block.custom

import com.mojang.serialization.MapCodec
import net.Chidoziealways.everythingjapanese.portal.HellPortalForcer
import net.Chidoziealways.everythingjapanese.portal.HellPortalShape
import net.Chidoziealways.everythingjapanese.worldgen.dimension.ModDimensions
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.server.level.ServerLevel
import net.minecraft.util.BlockUtil
import net.minecraft.util.RandomSource
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.InsideBlockEffectApplier
import net.minecraft.world.entity.Relative
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.BlockGetter
import net.minecraft.world.level.Level
import net.minecraft.world.level.LevelReader
import net.minecraft.world.level.ScheduledTickAccess
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.block.Portal
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.StateDefinition
import net.minecraft.world.level.block.state.properties.BlockStateProperties
import net.minecraft.world.level.block.state.properties.EnumProperty
import net.minecraft.world.level.border.WorldBorder
import net.minecraft.world.level.dimension.DimensionType
import net.minecraft.world.level.gamerules.GameRules
import net.minecraft.world.level.portal.PortalShape
import net.minecraft.world.level.portal.TeleportTransition
import net.minecraft.world.level.portal.TeleportTransition.PostTeleportTransition
import net.minecraft.world.phys.Vec3
import net.minecraft.world.phys.shapes.Shapes
import net.minecraft.world.phys.shapes.VoxelShape
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import kotlin.math.max

class HellPortalBlock(properties: Properties) : Block(properties), Portal {
    override fun codec(): MapCodec<out Block?> {
        return CODEC
    }

    init {
        this.registerDefaultState(
            this.stateDefinition.any().setValue<Direction.Axis, Direction.Axis>(AXIS, Direction.Axis.X)
        )
    }

    override fun getEntityInsideCollisionShape(
        blockState: BlockState,
        blockGetter: BlockGetter,
        blockPos: BlockPos,
        entity: Entity
    ): VoxelShape {
        return blockState.getShape(blockGetter, blockPos)
    }

    override fun entityInside(
        blockState: BlockState,
        level: Level,
        blockPos: BlockPos,
        entity: Entity,
        insideBlockEffectApplier: InsideBlockEffectApplier,
        boolean: Boolean
    ) {
        if (entity.canUsePortal(false)) {
            entity.setAsInsidePortal(this, blockPos)
        }
    }

    override fun updateShape(
        state: BlockState,
        level: LevelReader,
        scheduledTickAccess: ScheduledTickAccess,
        pos: BlockPos,
        direction: Direction,
        neighborPos: BlockPos,
        neighborState: BlockState,
        random: RandomSource
    ): BlockState {
        val axis = direction.axis
        val axis1 = state.getValue(AXIS)
        val flag = axis1 != axis && axis.isHorizontal
        return if(!flag && !neighborState.`is`(this) && !HellPortalShape.findAnyShape(level, pos, axis1).isComplete) Blocks.AIR.defaultBlockState() else super.updateShape(state, level, scheduledTickAccess, pos, direction, neighborPos, neighborState, random)
    }

    override fun getPortalTransitionTime(serverLevel: ServerLevel, entity: Entity): Int {
        return if (entity is Player) max(
            0,
            serverLevel.gameRules.get (if (entity.abilities.invulnerable)  GameRules.PLAYERS_NETHER_PORTAL_CREATIVE_DELAY else GameRules.PLAYERS_NETHER_PORTAL_DEFAULT_DELAY)
        ) else
            0
    }

    override fun getPortalDestination(pLevel: ServerLevel, pEntity: Entity, pPos: BlockPos): TeleportTransition? {
        val resourcekey =
            if (pLevel.dimension() === ModDimensions.HELL_LEVEL_KEY) Level.OVERWORLD else ModDimensions.HELL_LEVEL_KEY
        val serverlevel = pLevel.server.getLevel(resourcekey)
        if (serverlevel == null) {
            return null
        } else {
            val flag = serverlevel.dimension() === ModDimensions.HELL_LEVEL_KEY
            val worldborder = serverlevel.worldBorder
            val d0 = DimensionType.getTeleportationScale(pLevel.dimensionType(), serverlevel.dimensionType())
            val blockpos = worldborder.clampToBounds(pEntity.x * d0, pEntity.y, pEntity.z * d0)
            return this.getExitPortal(serverlevel, pEntity, pPos, blockpos, flag, worldborder)
        }
    }

    private fun getExitPortal(
        pLevel: ServerLevel,
        pEntity: Entity,
        pPos: BlockPos,
        pExitPos: BlockPos,
        pIsNether: Boolean,
        pWorldBorder: WorldBorder
    ): TeleportTransition? {
        val forcer = HellPortalForcer(pLevel)
        val optional = forcer.findClosestPortalPosition(pExitPos, pIsNether, pWorldBorder)
        val foundRectangle: BlockUtil.FoundRectangle
        val postTeleportTransition: PostTeleportTransition
        if (optional.isPresent) {
            val blockpos = optional.get()
            val blockstate = pLevel.getBlockState(blockpos)
            foundRectangle = BlockUtil.getLargestRectangleAround(
                blockpos,
                blockstate.getValue(BlockStateProperties.HORIZONTAL_AXIS),
                21,
                Direction.Axis.Y,
                21
            ) { pos: BlockPos -> pLevel.getBlockState(pos) === blockstate }
            postTeleportTransition =
                TeleportTransition.PLAY_PORTAL_SOUND.then { entity: Entity ->
                    entity.placePortalTicket(blockpos)
                }
        } else {
            val directionAxis = pEntity.level().getBlockState(pPos).getOptionalValue(AXIS).orElse(
                Direction.Axis.X
            )
            val optional1 = forcer.createPortal(pExitPos, directionAxis)
            if (optional1.isEmpty) {
                LOGGER.error("Unable to create a portal, likely target out of worldborder")
                return null
            }

            foundRectangle = optional1.get()
            postTeleportTransition =
                TeleportTransition.PLAY_PORTAL_SOUND.then(TeleportTransition.PLACE_PORTAL_TICKET)
        }

        return getDimensionTransitionFromExit(
            pEntity,
            pPos,
            foundRectangle,
            pLevel,
            postTeleportTransition
        )
    }

    override fun getLocalTransition(): Portal.Transition {
        return Portal.Transition.CONFUSION
    }

    @Deprecated("Deprecated in Java")
    override fun getCloneItemStack(
        pLevel: LevelReader,
        pPos: BlockPos,
        pState: BlockState,
        pIncludeData: Boolean
    ): ItemStack {
        return ItemStack.EMPTY
    }

    override fun createBlockStateDefinition(pBuilder: StateDefinition.Builder<Block, BlockState>) {
        pBuilder.add(AXIS)
    }

    companion object {
        private val LOGGER: Logger = LoggerFactory.getLogger(HellPortalBlock::class.java)
        val CODEC: MapCodec<HellPortalBlock> = simpleCodec<HellPortalBlock> { properties: Properties ->
            HellPortalBlock(
                properties
            )
        }
        val AXIS: EnumProperty<Direction.Axis> = BlockStateProperties.HORIZONTAL_AXIS
        private val SHAPES: MutableMap<Direction.Axis, VoxelShape> =
            Shapes.rotateHorizontalAxis(column(4.0, 16.0, 0.0, 16.0))

        private fun getDimensionTransitionFromExit(
            pEntity: Entity,
            pPos: BlockPos,
            pRectangle: BlockUtil.FoundRectangle,
            pLevel: ServerLevel,
            pPostTeleportTransition: PostTeleportTransition
        ): TeleportTransition {
            val blockstate = pEntity.level().getBlockState(pPos)
            val directionAxis: Direction.Axis?
            val vec3: Vec3?
            if (blockstate.hasProperty(BlockStateProperties.HORIZONTAL_AXIS)) {
                directionAxis = blockstate.getValue(BlockStateProperties.HORIZONTAL_AXIS)
                val blockUtilFoundRectangle = BlockUtil.getLargestRectangleAround(
                    pPos,
                    directionAxis,
                    21,
                    Direction.Axis.Y,
                    21
                ) { blockPos: BlockPos -> pEntity.level().getBlockState(blockPos) === blockstate }
                vec3 = pEntity.getRelativePortalPosition(directionAxis, blockUtilFoundRectangle)
            } else {
                directionAxis = Direction.Axis.X
                vec3 = Vec3(0.5, 0.0, 0.0)
            }

            return createDimensionTransition(
                pLevel,
                pRectangle,
                directionAxis,
                vec3,
                pEntity,
                pPostTeleportTransition
            )
        }

        private fun createDimensionTransition(
            pLevel: ServerLevel,
            pRectangle: BlockUtil.FoundRectangle,
            pAxis: Direction.Axis?,
            pOffset: Vec3,
            pEntity: Entity,
            pPostTeleportTransition: PostTeleportTransition
        ): TeleportTransition {
            val blockpos = pRectangle.minCorner
            val blockstate = pLevel.getBlockState(blockpos)
            val directionAxis =
                blockstate.getOptionalValue(BlockStateProperties.HORIZONTAL_AXIS).orElse(
                    Direction.Axis.X
                )
            val d0 = pRectangle.axis1Size.toDouble()
            val d1 = pRectangle.axis2Size.toDouble()
            val entitydimensions = pEntity.getDimensions(pEntity.pose)
            val i = if (pAxis === directionAxis) 0 else 90
            val d2 = entitydimensions.width() / 2.0 + (d0 - entitydimensions.width()) * pOffset.x()
            val d3 = (d1 - entitydimensions.height()) * pOffset.y()
            val d4 = 0.5 + pOffset.z()
            val flag = directionAxis === Direction.Axis.X
            val vec3 = Vec3(
                blockpos.x + (if (flag) d2 else d4),
                blockpos.y + d3,
                blockpos.z + (if (flag) d4 else d2)
            )
            val vec31 = PortalShape.findCollisionFreePosition(vec3, pLevel, pEntity, entitydimensions)
            return TeleportTransition(
                pLevel,
                vec31,
                Vec3.ZERO,
                i.toFloat(),
                0.0f,
                Relative.union(Relative.DELTA, Relative.ROTATION),
                pPostTeleportTransition
            )
        }
    }
}
